test/workflow-orchestration.ts

```diff
@@ -138,6 +138,70 @@ describe('Workflow chaining for a collection configured for swot reprojection an
+    const reprojectAndZarrQuery = {
+      maxResults: 3,
+      outputCrs: 'EPSG:4326',
+      interpolation: 'near',
+      scaleExtent: '0,2500000.3,1500000,3300000',
+      scaleSize: '1.1,2',
+      format: 'application/x-zarr',
+      turbo: true,
+    };
+
+    hookRangesetRequest('1.0.0', collection, 'all', { query: reprojectAndZarrQuery });
+    hookRedirect('joe');
+
+    describe('when checking for a query-cmr work item', function () {
+      it('finds and completes it, queueing the next work item for each result', async function () {
+        const res = await getWorkForService(this.backend, 'harmonyservices/query-cmr:latest');
+        expect(res.status).to.equal(200);
+        const workItem = JSON.parse(res.text);
+        expect(workItem.serviceID).to.equal('harmonyservices/query-cmr:latest');
+        workItem.status = WorkItemStatus.SUCCESSFUL;
+        workItem.results = [
+          'test/resources/worker-response-sample/catalog0.json',
+          'test/resources/worker-response-sample/catalog1.json',
+          'test/resources/worker-response-sample/catalog2.json',
+        ];
+        await updateWorkItem(this.backend, workItem);
+        // since there were multiple query cmr results,
+        // multiple work items should be generated for the next step
+        const currentWorkItems = (await getWorkItemsByJobId(db, workItem.jobID)).workItems;
+        expect(currentWorkItems.length).to.equal(4);
+        expect(currentWorkItems.filter((i) => i.status === WorkItemStatus.READY && i.serviceID === 'sds/swot-reproject:latest').length).to.equal(3);
+      });
+
+      describe('when the first swot-reprojection service work item fails', function () {
+        it('the job fails, and all further work items are canceled', async function () {
+          const res = await getWorkForService(this.backend, 'sds/swot-reproject:latest');
+          expect(res.status).to.equal(200);
+          const workItem = JSON.parse(res.text);
+          workItem.status = WorkItemStatus.FAILED;
+          workItem.results = [];
+          await updateWorkItem(this.backend, workItem);
+          expect(workItem.serviceID).to.equal('sds/swot-reproject:latest');
+          // work item failure should trigger job failure
+          const job = await Job.byJobID(db, workItem.jobID);
+          expect(job.status === JobStatus.FAILED);
+          // job failure should trigger cancellation of any pending work items
+          const currentWorkItems = (await getWorkItemsByJobId(db, workItem.jobID)).workItems;
+          expect(currentWorkItems.length).to.equal(4);
+          expect(currentWorkItems.filter((i) => i.status === WorkItemStatus.SUCCESSFUL && i.serviceID === 'harmonyservices/query-cmr:latest').length).to.equal(1);
+          expect(currentWorkItems.filter((i) => i.status === WorkItemStatus.CANCELED && i.serviceID === 'sds/swot-reproject:latest').length).to.equal(2);
```

Comment:
One way to guard against the issue I mentioned earlier, where we could add a status and it ends up not getting handled correctly, would be to loop over the statuses in the `WorkItemStatus` enum and ensure that one item per status exists, then the test could assert that all statuses end up in the correct bucket.